### PR TITLE
Write NU1004 errors to the assets file, ensure that all 3 restore output files are generated in all failure scenarios

### LINKED ISSUE #8973: Error NU1004 disappears when building solution after enabling “RestoreLockedMode”
## Details about Problem

NuGet Version: Dev\5.5.0.6330
VS Version: Master\29623.15
OS Version: Windows-10-Enterprise-RS5

## Installation steps: 
* https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3341765&view=results

## Repro steps: 
1. Create above corresponding projects via VS. 
2. Install a few ‘PackageReference’ packages into the projects.
3. Set MSBuild property “RestorePackagesWithLockFile” to true in csproj file.
  `<RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>`
4. Do a restore.
5. Check it should have generated ‘packages.lock.json’ file alongside *.csproj file.
6. Set msbuild property “RestoreLockedMode” to “true”.
    `<RestoreLockedMode>true</RestoreLockedMode>`
7. Install another package.
8. It should still work fine and update ‘packages.lock.json’ file.
9. Manually add a new PackageReference item into csproj file.
```
<PackageReference Include="NUnit">
   <Version>3.12.0</Version>
 </PackageReference>
```
10. Perform restore.
11. Restore should fail with error “NU1004”
12. Build the solution.

## Expected:
The error was still showed on error list.

## Actual:
The error disappear on error list.

## Here is the record video for the repro steps:
restore

## Notes: 
1. The issue repro on .NET Core project.
2. The issue repro on Master\29623.15 + NuGet Client Dev\5.5.0.6355.
3. The issue is not stable, the repro rate is almost 95% for us.

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/8973
Regression: No 
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 

The restore command collects warnings and errors using a `CollectorLogger` that's initialized from the logger passed in to the restore command. 
https://github.com/NuGet/NuGet.Client/blob/85f17f0525048c36cdd080324019729884b1d998/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs#L91-L95

When restore fails/succeeds and the assets file is generated, the collector logger is inspected to generated the correct warnings/errors. All warnings and errors are written to the assets file.

NU1004 was incorrectly using the non collector logger so those errors were never considered. 
Furthermore when NU1004 was raised, the old assets file was provided to the result, thus one that would not contain the log messages. 

## Testing/Validation

Tests Added: Yes 
Reason for not adding tests:  
Validation:

### CODE REVIEW

**Comment by nkolev92** [comment]
This fix broke a different scenario. 
This PR still needs some work.

**Comment by nkolev92** [comment]
Ready for review again.

**Comment by nkolev92** [comment]
🔔 @NuGet/nuget-client

**Review by kartheekp-ms — APPROVED** [review]

File: src/NuGet.Core/NuGet.Commands/Strings.resx [hunk 1]
```diff
@@ -1118,4 +1118,7 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
     <value>Platform version is not present for one or more target frameworks, even though they have specified a platform: {0}</value>
     <comment>0 - comma-separated list of TFMs missing a platform version</comment>
   </data>
+  <data name="Log_SkippingPackagesLockFileGeneration" xml:space="preserve">
+    <value>Skipping the lock file regeneration for '{0}'.</value>
```
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Commands/Strings.resx:1122]: Don't you need a comment to define {0} for localization? Also, I just had a case where they didn't receive any of my comments... :(

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs [hunk 2]
```diff
@@ -1175,6 +1156,9 @@ private static RemoteWalkContext CreateRemoteWalkContext(RestoreRequest request,
                 context.RemoteLibraryProviders.Add(provider);
             }
 
+            // Determine if the targets and props files should be written out.
+            context.IsMsBuildBased = request.ProjectStyle != ProjectStyle.DotnetCliTool;
```
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:1155]: So `Unknown` and `ProjectJson` and `DotnetToolReference` are MsBuildBased?

Assuming the property is already set and you're just excluding one option...
```suggestion
            context.IsMsBuildBased = context.IsMsBuildBased && request.ProjectStyle != ProjectStyle.DotnetCliTool;
```

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs [hunk 3]
```diff
@@ -595,7 +579,7 @@ private bool ValidatePackagesSha512(PackagesLockFile lockFile, LockFile assetsFi
                         // bail restore since it's the locked mode but required to update the lock file.
                         // directly log to the request logger when we're not going to rewrite the assets file otherwise this log will
                         // be skipped for netcore projects.
-                        await _request.Log.LogAsync(RestoreLogMessage.CreateError(NuGetLogCode.NU1004, Strings.Error_RestoreInLockedMode));
+                        await _logger.LogAsync(RestoreLogMessage.CreateError(NuGetLogCode.NU1004, Strings.Error_RestoreInLockedMode));
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:577]: Here you changed the logger, but line#580 it states `// directly log to the request logger when we're not going to rewrite the assets file otherwise this log will`.
So now is this _logger (instance of RestoreCollectorLogger) actually writing anything to lock file or asset file?

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs [hunk 4]
```diff
@@ -870,6 +872,70 @@ public async Task Restore_LegacyPackageReference_P2P_SkipBuildTransitive()
...
+                    Version = "1.0.0"
+                };
+                packageX.Files.Clear();
+                packageX.AddFile("lib/net461/a.dll");
+
+                var packageY = new SimpleTestPackageContext()
+                {
+                    Id = "y",
+                    Version = "1.0.0"
+                };
```
Comment (erdembayar) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs:903]: You cleared files for `packageX` few lines above here. But why you're not doing it for `packageY`?

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs [hunk 5]
```diff
@@ -870,6 +872,70 @@ public async Task Restore_LegacyPackageReference_P2P_SkipBuildTransitive()
...
+
+                await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                    pathContext.PackageSource,
+                    PackageSaveMode.Defaultv3,
+                    packageX,
+                    packageY);
+
+                var result = RunRestore(pathContext, _successExitCode, "-UseLockFile");
+                Assert.True(result.Success);
+                Assert.True(File.Exists(projectA.NuGetLockFileOutputPath));
```
Comment (erdembayar) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs:919]: Do we need to check if lock file touched by add `packageY` action later?
If yes then we need to keep content hash or `last_modified` date and compare after adding `packageY`.

**Review by erdembayar — COMMENTED** [review]
In general looks good for me. Just few comment questions since it's my first code review on lock file.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:1155]: If you look at the diff more closely, this is not a functional change, it's just a timing change. 
The context property gets initialized at an earlier/more appropriate point.

File: src/NuGet.Core/NuGet.Commands/Strings.resx (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/Strings.resx:1122]: No, not really. The comment is from the localization team/engineering team, it has no effect on the formatting.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:577]: nice catch. The comment is outdated and irrelevant. I thought I had removed it. 

the _logger is a collector logger, so it will overwrite the assets file, which is necessary for the log message to appear in the error list.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs (hunk 4 above)
Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs:903]: oversight. The defaults are net45 so in this case it was inconsequential. I added the clear for Y as well.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs (hunk 5 above)
Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs:919]: Not sure which content hash you are suggesting here, but it does make sense to check for the timestamp of the lock file. 

LockedMode *must not* overwrite the lock file. 
I added a timestamp check.

**Review by nkolev92 — COMMENTED** [review]
Addressed feedback.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs [hunk 6]
```diff
@@ -870,6 +872,75 @@ public async Task Restore_LegacyPackageReference_P2P_SkipBuildTransitive()
...
+                var originalPackagesLockFileWriteTime = new FileInfo(projectA.NuGetLockFileOutputPath).LastWriteTimeUtc;
+
+                projectA.AddPackageToAllFrameworks(packageY);
+                projectA.Save();
+
+                // Remove old obj folders.
+                Directory.Delete(Path.GetDirectoryName(projectA.AssetsFileOutputPath), recursive: true);
+
+
```
Comment (erdembayar) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs]: Delete 1 empty line here. Now there are 2 lines.

**Review by erdembayar — APPROVED** [review]
Please check 1 extra line, otherwise looks good.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs (hunk 6 above)
Comment (nkolev92) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs]: fixed.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs [hunk 7]
```diff
@@ -200,35 +201,8 @@ public async Task<RestoreResult> ExecuteAsync(CancellationToken token)
...
-                            cacheFile: cacheFile,
-                            cacheFilePath: _request.Project.RestoreMetadata.CacheFilePath,
-                            packagesLockFilePath: packagesLockFilePath,
-                            packagesLockFile: packagesLockFile,
-                            dependencyGraphSpecFilePath: NoOpRestoreUtilities.GetPersistedDGSpecFilePath(_request),
-                            dependencyGraphSpec: _request.DependencyGraphSpec,
-                            projectStyle: _request.ProjectStyle,
-                            elapsedTime: restoreTime.Elapsed);
-                    }
+                    regenerateLockFile = result; // Ensure that the lock file *does not* get rewritten, when the lock file is out of date and the status is false.
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:204]: I am bit confused with this change since I last reviewed. Looked into `EvaluatePackagesLockFileAsync` method implementation. `result` can be true even though `lockfile is not valid` but the reverse is `not true` which means if the `lock file is valid` then `result` cannot be `false`. 

I could be wrong but `EvaluatePackagesLockFileAsync` method returns `true` when `lock file is valid` then the current logic generates lockfile again.  Happy to learn more if this comment is incorrect.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:204]: > EvaluatePackagesLockFileAsync method implementation. result can be true even though lockfile is not valid

Yes, but there's more nuance to it. 

True/False can happen in 2 cases - The lock file is not enabled or the lock exists but the force evaluate flag was used.  

If you look here: https://github.com/NuGet/NuGet.Client/blob/11621e13bfcd1cbb8336bdaffcc1f180e9dc4350/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs#L307-L331, you'll see that the isLockFileValid check to prevent unnecessary generation. 

If you look at the description it now says `isLockFileValid tells whether lock file is still valid to be consumed for this restore`, so basically you can think of this as is lock file up to date. 

> but the reverse is not true which means if the lock file is valid then result cannot be false.

That is correct. When the lock file is valid to be used, that means we're either in locked mode or unlocked mode and the file is up to date, so success can't be false.

**Comment by nkolev92** [comment]
@donnie-msft 

You planning on reviewing again or can I merge?

File: src/NuGet.Core/NuGet.Commands/Strings.resx (hunk 1 above)
Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Commands/Strings.resx:1122]: Let me rephrase this... the localization team needs to know what your {0} means. I don't see where you have defined it using a comment.

**Comment by donnie-msft** [comment]
@nkolev92 I didn't block... but did reply to one of your replies.

File: src/NuGet.Core/NuGet.Commands/Strings.resx (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/Strings.resx:1122]: `the localization team needs to know what your {0} means.`

Wasn't aware of that. Many of the strings in this file do not have a comment, so I've never though otherwise. 
Especially with sentences that have only 1 format prop.
